### PR TITLE
222-Add-graph-manager

### LINKED ISSUE #222: Add graph/number manager for each detector mesh view
Each mesh view can have mulitple item.

If mesh is zero-dimensional (i.e. user chosen zone as a mesh) then we need to present a series of numbers.
This can be done in a table, with columns like:
- Name of quantity and unit
- list of names of filters
- calculated value
- optional calculated error

If mesh is 1D or 2D we need to present a series of canvases (many) with plots.

### PR DESCRIPTION
**Pull request by retkiewi**
Fixes #222

### CODE REVIEW

File: src/JsRoot/JsRootGraph2D.tsx [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+		return () => setIsVisible(false);
+	}, [visible]);
+
+	useEffect(() => {
+		if (drawn || !visible) return;
+        const x = props.first_axis.values;
+        const y = props.second_axis.values;
+        const z = props.data.values
+		
+		// JSRoot is unable to draw bins with value from the end of the range unless there is at least
```
Comment (grzanka) [inline src/JsRoot/JsRootGraph2D.tsx:26]: It is not a problem and rather a feature, I am quite happy with white bins for which value is zero.

File: src/JsRoot/JsRootGraph2D.tsx [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        gr.fXaxis.fXmin = x[0]
+        gr.fXaxis.fXmax = x[nxpoints-1]
+		gr.fXaxis.fTitle = `${props.first_axis.name} [${props.first_axis.unit}]`
+
+        gr.fYaxis.fXmin = y[0]
+        gr.fYaxis.fXmax = y[nypoints-1]
+		gr.fYaxis.fTitle = `${props.second_axis.name} [${props.second_axis.unit}]`
+
+        gr.fTitle = `${props.data.name} [${props.data.unit}]`;
+
```
Comment (grzanka) [inline src/JsRoot/JsRootGraph2D.tsx:45]: Fix indendation or use some autoformatter

File: src/WrapperApp/components/ResultsPanel.tsx [hunk 3]
```diff
@@ -1,37 +1,44 @@
...
+// 				.map((v, i) => {
+// 					const rand = Math.round(Math.random() * numberOfElements) + 1;
+// 					return { uuid: i, title: `Mesh${index} - Graph${i} - ${rand}`, value: rand };
+// 				})
+// 		};
+// 		data.push(obj);
+// 	}
+
+// 	return data;
+// })();
```
Comment (grzanka) [inline src/WrapperApp/components/ResultsPanel.tsx:30]:
```suggestion
```

is it needed ?

File: src/WrapperApp/components/ResultsPanel.tsx [hunk 4]
```diff
@@ -1,37 +1,43 @@
...
+// 			name: `Scoring Mesh${index}`,
+// 			graphs: Array(numberOfElements)
+// 				.fill(0)
+// 				.map((v, i) => {
+// 					const rand = Math.round(Math.random() * numberOfElements) + 1;
+// 					return { uuid: i, title: `Mesh${index} - Graph${i} - ${rand}`, value: rand };
+// 				})
+// 		};
+// 		data.push(obj);
+// 	}
```
Comment (grzanka) [inline src/WrapperApp/components/ResultsPanel.tsx:27]:
```suggestion
```

File: src/WrapperApp/components/ResultsPanel.tsx [hunk 5]
```diff
@@ -1,37 +1,43 @@
...
 
-	return data;
-})();
+// 	return data;
+// })();
```
Comment (grzanka) [inline src/WrapperApp/components/ResultsPanel.tsx:30]:
```suggestion
```

**Comment by grzanka** [comment]
@retkiewi please resolve conflicts  and remove hardcoded large JSON blob

File: src/JsRoot/JsRootGraph2D.tsx [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+export function JsRootGraph2D(props: Page2D) {
+	const { JSROOT } = useJSROOT();
+	const [containerEl, visible] = useVisible<HTMLDivElement>();
+
+	const [obj, setObj] = useState(undefined);
+	const [drawn, setDrawn] = useState(false);
+	const [isVisible, setIsVisible] = useState(false);
+
+	useEffect(() => {
+		setIsVisible(visible > 0.3);
```
Comment (grzanka) [inline src/JsRoot/JsRootGraph2D.tsx:16]: what is that 0.3 constant ?

File: src/JsRoot/JsRootGraph2D.tsx [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+		gr.fYaxis.fXmax = y[nypoints - 1];
+		gr.fYaxis.fTitle = `${props.second_axis.name} [${props.second_axis.unit}]`;
+
+		gr.fTitle = `${props.data.name} [${props.data.unit}]`;
+
+		for (let x = 0; x < nxpoints; x++)
+			for (let y = 0; y < nypoints; y++) {
+				gr.setBinContent(gr.getBin(x + 1, y + 1), z[x + y * nxpoints]);
+			}
+
```
Comment (grzanka) [inline src/JsRoot/JsRootGraph2D.tsx:45]: This looks like local `x`,`y` are shadowing the same variable defined above:

```
	const x = props.first_axis.values;
	const y = props.second_axis.values;
```

Comment (grzanka) [reply src/JsRoot/JsRootGraph2D.tsx:45]: Also - explain why do you need to use `+1` in `getBin` method

File: src/JsRoot/JsRootGraph2D.tsx [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+		setObj(gr);
+		setDrawn(false);
+	}, [JSROOT, props, drawn, visible]);
+
+	useEffect(() => {
+		if (obj && !drawn && isVisible) {
+			JSROOT.cleanup(containerEl.current);
+			JSROOT.redraw(containerEl.current, obj, 'col');
```
Comment (grzanka) [inline src/JsRoot/JsRootGraph2D.tsx:53]: Can you try `COLZ` ?
According to
https://root.cern.ch/doc/v608/classTHistPainter.html

```
Same as "COL". In addition the color palette is also drawn. 
```

File: src/JsRoot/JsRootGraph2D.tsx [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+	useEffect(() => {
+		if (drawn || !visible) return;
+		const x = props.first_axis.values;
+		const y = props.second_axis.values;
+		const z = props.data.values;
+
+		const nxpoints = x.length;
+		const nypoints = y.length;
+
+		const gr = JSROOT.createHistogram('TH2F', nxpoints, nypoints);
```
Comment (grzanka) [inline src/JsRoot/JsRootGraph2D.tsx:29]: what is the meaning of `gr` variable name ?

File: src/JsRoot/JsRootGraph2D.tsx (hunk 6 above)
Comment (retkiewi) [reply src/JsRoot/JsRootGraph2D.tsx:16]: It's an arbitrary constant that determines if the graph should be loaded based on how much of the graph is visible.

Comment (grzanka) [reply src/JsRoot/JsRootGraph2D.tsx:19]: ok

**Review by grzanka — APPROVED** [review]
LGTM

**Pull request merged**